clang/test/CodeGenSYCL/device_global.cpp

```diff
@@ -0,0 +1,86 @@
+// RUN: %clang_cc1 -fsycl-is-device -internal-isystem %S/Inputs -triple spir64-unknown-unknown-sycldevice -disable-llvm-passes -fsycl-unique-prefix=THE_PREFIX -std=c++17 -emit-llvm %s -o - | FileCheck %s
+#include "sycl.hpp"
+
+// Test cases below show that 'sycl-unique-id' LLVM IR attribute is attached to the
+// global variable whose type is decorated with device_global attribute, and that a
+// unique string is generated.
+//
+namespace cl {
+namespace sycl {
+namespace ext {
+namespace oneapi {
+// decorated with only global_variable_allowed attribute
+template <typename T>
+class [[__sycl_detail__::global_variable_allowed]] only_global_var_allowed {
+public:
+  const T &get() const noexcept { return *Data; }
+  only_global_var_allowed() {}
+  operator T &() noexcept { return *Data; }
+
+private:
+  T *Data;
+};
+} // namespace oneapi
+} // namespace ext
+} // namespace sycl
+} // namespace cl
+
+using namespace sycl::ext::oneapi;
+using namespace cl::sycl;
+queue q;
+
+device_global<int> A;
+static device_global<int> B;
+
+struct Foo {
+  static device_global<int> C;
+};
+device_global<int> Foo::C;
+// CHECK: @A = addrspace(1) global %"class.cl::sycl::ext::oneapi::device_global" zeroinitializer, align 8 #[[A_ATTRS:[0-9]+]]
+// CHECK: @_ZL1B = internal addrspace(1) global %"class.cl::sycl::ext::oneapi::device_global" zeroinitializer, align 8 #[[B_ATTRS:[0-9]+]]
+// CHECK: @_ZN3Foo1CE = addrspace(1) global %"class.cl::sycl::ext::oneapi::device_global" zeroinitializer, align 8 #[[C_ATTRS:[0-9]+]]
+
+device_global<int> same_name;
+namespace NS {
+device_global<int> same_name;
+}
+// CHECK: @same_name = addrspace(1) global %"class.cl::sycl::ext::oneapi::device_global" zeroinitializer, align 8 #[[SAME_NAME_ATTRS:[0-9]+]]
+// CHECK: @_ZN2NS9same_nameE = addrspace(1) global %"class.cl::sycl::ext::oneapi::device_global" zeroinitializer, align 8 #[[SAME_NAME_NS_ATTRS:[0-9]+]]
+
+// check that we don't generate `sycl-unique-id` IR attribute if class does not use
+// [[__sycl_detail__::device_global]]
+only_global_var_allowed<int> no_device_global;
```

Comment:
I would add a couple of more test cases, for example with template:
```
template <typename T> struct TS {    
public:                          
  static device_global<T> d;             
}; 
...
// and then use inside kernel like others
auto AA = TS::d<int>.get();
```
With inline namespace:
```
inline namespace Bar {
device_global<float> InlineNS;
}
... 
// use InlineNS from the kernel as well
```
Something like
```
template <typename T>
const device_global<T> const_dev_global;
// And then in kernel
auto val = const_dev_global<int>.get();
```

Reply:
For the first case, using it like this inside the kernel appears syntactically correct: 
` (void)TS<int>::d.get();
`
However, you are asking for a case with d<int>, correct?

The third case doesn't work either - Will fix it in the next PR.

Reply:
https://github.com/intel/llvm/pull/5597/commits/5bced1a651bb78c46754af0c26a75a006950232f